students/Ruohan/lesson09/mailroom_oo_new.py

```diff
@@ -0,0 +1,173 @@
+            if not min_d is None and max_d is None:
+                donor.donation = list(filter(lambda x: x >= min_d and x <= max_d, donor.donation))
+            else:
+                continue
+            donor.donation = map(lambda x : x * n, donor.donation)
+        return self._donors
+
+def Welcome():
+    print('Welcome to Mailroom')
+    db = donor_DB()
+    while True:
+        print('\nwhat do you want to do?')
+        print('1) Check the donor list\n'
+              '2) A new donation\n'
+              '3) Send thank you\n'
+              '4) Creat a report\n'
+              '5) Send letters to everyone\n'
+              '6) Total contribution by MORE\n'
+              '7) quit\n')
+        response = input('>> ')
+        if response == '1':
+            print(db.donor_list())
+            print('Return to main menu\n')
+        elif response == '2':
+            name = input('Enter the donor name: ').lower()
+            db.add_donor(name)
+            amount = input('Enter the donation amount: ')
+            try: amount = float(amount)
+            except ValueError:
+                print('error: donation amount is invalid\n')
+            db.search_donor(name).add_donations(amount)
+            print('New donation has been recorded')
+            print('Return to main menu\n')
+        elif response == '3':
+            name = input('Who do you want to send thank you letter to: ').lower()
+            if db.find_donor(name):
+                print(db._letter(db.find_donor(name)))
+            else:
+                print('Donor is not recorded, please add donor to datasets first(chose 2 in main menu)\n')
+                print('Return to main menu\n')
+                continue
+        elif response == '4':
+            for line in db.report():
+                print(line)
+            print('Return to main menu\n')
+        elif response == '5':
+            db.file_letters()
+            print('Letters have been saved')
+            print('Return to main menu\n')
+        elif response == '6':
+            print('scenario 1: double contributions under $100\n'
+                  'scenario 2: triple contributions over $50\n')
+            rs = input('choose your scenario "1" or "2": ')
+            if rs == '1':
+                n = 2
+                max = 100
+                donor_new_data = db.challenge(n, max_d = max)
+            elif rs == '2':
+                n = 3
+                min = 50
+                donor_new_data = db.challenge(n, min_d = min)
+            else:
+                print('Invalid input')
+            for donor in donor_new_data.values():
+                print(donor.name,' new total donation is ', reduce(lambda x, y: x + y, donor.donation))
+        elif response == '7':
+            print('Exit Now')
+            db.quit()
+        else:
+            print('Invalid input')
+            print('Return to main menu\n')
+            continue
+
+
```

Comment:
This looks great, I guess this is the one you went over with Hosung, so let us know if you have other questions, but this seems to work